### DISCUSSION TITLE
Ts-node module resolution is not respected

### DISCUSSION THREAD

**Discussion by lantos1618** [TypeStrong/ts-node#1859] [General]
### Search Terms
ts-node module type resolution package.json
TypeError [ERR_UNKNOWN_FILE_EXTENSION]: Unknown file extension ".ts" for /...

### Expected Behavior
when "moduleType" is specified in ts-config.json it should ignore the "type": "module" specified in package.json

tsconfig.json
```json
  "ts-node": {
    "moduleTypes": {
      "*": "commonjs"
    }
  },
```
package.json
```
{
  "name": "typescript-incorrect-behaviour",
  "version": "1.0.0",
  "description": "",
  "type": "module",
  "main": "index.js",
```

### Actual Behavior

  "type": "module", is respected when tsconfig.json should be respected.

Even passing "NODE_OPTIONS='--import-type=commonjs'" does not get respected.

throws
```
Debugger attached.
TypeError [ERR_UNKNOWN_FILE_EXTENSION]: Unknown file extension ".ts" for ~/Desktop/typescript-incorrect-behaviour/src/index.ts
    at new NodeError (node:internal/errors:372:5)
    at Object.getFileProtocolModuleFormat [as file:] (node:internal/modules/esm/get_format:76:11)
    at defaultGetFormat (node:internal/modules/esm/get_format:118:38)
    at defaultLoad (node:internal/modules/esm/load:21:20)
    at ESMLoader.load (node:internal/modules/esm/loader:407:26)
    at ESMLoader.moduleProvider (node:internal/modules/esm/loader:326:22)
    at new ModuleJob (node:internal/modules/esm/module_job:66:26)
    at ESMLoader.#createModuleJob (node:internal/modules/esm/loader:345:17)
    at ESMLoader.getModuleJob (node:internal/modules/esm/loader:304:34)
    at processTicksAndRejections (node:internal/process/task_queues:96:5) {
  code: 'ERR_UNKNOWN_FILE_EXTENSION'
}
```

### Steps to reproduce the problem
npm init
npm install typescript ts-node @types/node
tsc --init
create a src/index.ts with console.log("hello") inside
add "type": "module" to package.json
add "start": "ts-node src/index.ts" to package.json
npm start

### Minimal reproduction
I have an example here.
https://github.com/lantos-lgtm/typescript-incorrect-behaviour

```bash 
git clone  https://github.com/lantos-lgtm/typescript-incorrect-behaviour
cd typescript-incorrect-behaviour
npm install
npm start
```

### Specifications

ts-node v10.9.1
node v16.16.0
compiler v4.7.4
* tsconfig.json, if you're using one:
```
{
  "ts-node": {
    "moduleTypes": {
      "*": "commonjs"
    }
  },
  "compilerOptions": {
    "target": "es2016",
    "module": "commonjs",
    "esModuleInterop": true,
    "forceConsistentCasingInFileNames": true,
    "strict": true,
    "skipLibCheck": true
  }
}
```
* package.json:
```
{
  "name": "typescript-incorrect-behaviour",
  "version": "1.0.0",
  "description": "",
  "type": "module",
  "main": "index.js",
  "scripts": {
    "start": "ts-node src/index.ts",
    "start_node_env_override": "ts-node NODE_OPTIONS='--import-type=commonjs' src/index.ts",
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "dependencies": {
    "@types/node": "*",
    "ts-node": "^10.9.1",
    "typescript": "^4.7.4"
  }
}

```
* Operating system and version: macos 12.5
* If Windows, are you using WSL or WSL2?: none

**Comment by cspotcode** [comment]
I double-checked the docs, looks like moduleTypes might be configured incorrectly:
https://typestrong.org/ts-node/docs/module-type-overrides

**Comment by lantos1618** [comment]
Just updated to cjs still same response

**Reply by cspotcode** [reply]
Converted to discussion since this seems to be a support request, not a bug report.  I don't have bandwidth to answer support questions, so I recommend asking on the TypeScript Community Discord.

**Reply by lantos1618** [reply]
Thanks for spending time on this.
I don't think this is exactly user error. I had tried the "cjs" rather then "commonjs" before I was trying different combinations and just had committed the wrong version. 
I was going off https://typestrong.org/ts-node/docs/module-type-overrides

I'll give the discord a ping.

**Reply by lantos1618** [reply]
Potential solution for anyone who ends up in the same problem

cjs.tsconfig.json
```json
{
  "ts-node": {
    "moduleTypes": {
      "src/**/*.ts": "cjs"
    }
  },  
  "compilerOptions": {
    "target": "es2016",
    "module": "commonjs",
    "rootDir": ".",
    "esModuleInterop": true,
    "forceConsistentCasingInFileNames": true,
    "strict": true,
    "skipLibCheck": true
  }
}
```

package.json
```json
{
  "name": "my-app",
  "private": true,
  "version": "0.0.0",
  "type": "module",
  ... 
}
```

create a index.cjs file and require the index.ts
```js
require("index")
```

then you can run 
```
ts-node --project=cjs.tsconfig.json src/index.cjs
```

I had no luck trying "module": "esnext" and index.cts file

**Reply by lantos1618** [reply]
Or by

```json
  "ts-node": {
    "moduleTypes": {
      "src/**/*.ts": "cjs"
    },
    "esm": true,
 }
```

**Reply by cspotcode** [reply]
I think you had two problems: First, the `"commonjs"` should have been `"cjs"`.  Second, the glob pattern `"*"` was wrong.

**Reply by callmeberzerker** [reply]
I am bumping my head against the same issue. Did you found a more idiomatic solution @lantos-lgtm ? I really don't wanna do the `index.cjs` file approach as above, but whatever I do I can't make `moduleTypes` to be respected as `cjs`.

**Reply by lantos1618** [reply]
idiomatic, no.
hacky, yes.

I have two tsconfigs now and then specify which config to use using package.json

```json
    "server": "nodemon --project=tsconfig.server.json src/server/server.ts",
    "frontend": "nodemon --project=tsconfig.frontend.json src/index.ts"
```

I'm probably wrong but "nothing about ts is idiomatic." (joke). I saw you are doing a startup, I am doing the same. I am kinda deep into it change but looking back I personally would not choose TS for mid/long term development before you really consider what is important to you.  My cofounder and I spent way to long on things like this when we should have been actually writing our business logic...

**Reply by callmeberzerker** [reply]
No worries I never ever wanna go back to the era before `TypeScript` -> these quirks and ecosystem issues included. Yeah TS/JS has a wonky ecosystem. And ESM specifically is such a wonky thing in NodeJS even if you exclude TS.

P.S. In the end I managed to get it working `idiomatically`. Here is my `tsconfig.json` configuration for the `ts-node` property:

```json
{      
        "compilerOptions": {
		// stuff here
	},,
	"ts-node": {
		"moduleTypes": {
			"**/*.ts": "cjs"
		},
		"transpileOnly": true,
		"preferTsExts": true,
		"esm": false,
		"require": ["tsconfig-paths/register"],
		"compilerOptions": {
			"preserveValueImports": false
		}
	}
}
```
